ecommerce/serializers.py

```diff
@@ -116,3 +151,79 @@ def get_coupons(self, instance):
     class Meta:
         fields = ["items", "coupons"]
         model = models.Basket
+
+
+class CouponPaymentSerializer(serializers.ModelSerializer):
+    """ Serializer for coupon payments """
+
+    class Meta:
+        fields = "__all__"
+        model = models.CouponPayment
+
+
+class CouponPaymentVersionSerializer(serializers.ModelSerializer):
+    """ Serializer for coupon payment versions """
+
+    payment = CouponPaymentSerializer()
+
+    class Meta:
+        fields = "__all__"
+        model = models.CouponPaymentVersion
+
+
+class BaseCouponSerializer(serializers.Serializer):
+    """ Base serializer for coupon creation data """
+
+    name = serializers.CharField(
+        max_length=256,
+        validators=[UniqueValidator(queryset=CouponPayment.objects.all())],
+    )
+    tag = serializers.CharField(max_length=256, allow_null=True, required=False)
+    amount = serializers.DecimalField(decimal_places=2, max_digits=20)
+    automatic = serializers.BooleanField(default=False)
+    activation_date = DateTimeTzField()
+    expiration_date = DateTimeTzField()
+    products = serializers.ListField(child=IntegerField())
+    max_redemptions = serializers.IntegerField(default=1)
+    max_redemptions_per_user = serializers.IntegerField(default=1)
+    coupon_type = serializers.ChoiceField(
+        choices=[(_type, _type) for _type in CouponPaymentVersion.COUPON_TYPES]
+    )
+    num_coupon_codes = serializers.IntegerField(default=1)
+    company = serializers.CharField(max_length=512, allow_null=True, required=False)
+
+    def validate_products(self, value):
+        """ Determine if the products field is valid """
+        if not value:
+            raise ValidationError("At least one product must be selected")
+        for product_id in value:
+            if Product.objects.filter(id=product_id).exists() is False:
```

Comment:
It might be better to select `Product.objects.filter(id__in=value).values_list("id", flat=True)` and then raise an error if there's a `set()` difference. It'll save on queries a little and the error message will be comprehensive for the input.

Reply:
👍